### PR TITLE
[SDPA-4724] Tide anti-concurrent editing

### PR DESCRIPTION
**Pull request by vincent-gao**
### Jira
https://digital-engagement.atlassian.net/browse/SDPA-4724

### Features
1. Alert a user if they are navigating away from a modified form without saving.
   - admin user could define which form will apply unsaved warning to. 
   - support form: bundles , entity types, custom forms and regex supported. 
   - support user suddenly closes the browser's tab.
   - configurations could be exported.
 2. Preventing content editors editing a page at the same time.
   - support `tide_site_restriction` so that user can only view their own locked content
   - when a user is editing a node, any other user that attempts to edit the same node will be blocked from doing so, and notified that the content is already being edited.
   
### Screenshots
image
image
image
image
image

### CODE REVIEW

File: modules/tide_edit_protection/composer.json [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+    "name": "dpc-sdp/tide_edit_protection",
+    "description": "Prevents the user leaves the page without saving.",
```
Comment (anthony-malkoun) [inline modules/tide_edit_protection/composer.json:3]: "leaving"

File: modules/tide_edit_protection/src/Form/TideEditProtectionForm.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      'entity_type' => $this->t('Forms by entity type'),
+      'entity_bundle' => $this->t('Forms by entity bundle'),
+    ];
+
+    $form['entity_method'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Method'),
+      '#default_value' => $config->get('entity_method'),
+      '#options' => $entity_methods,
+      '#description' => $this->t('Choose which entity forms will be match for unsaved warning when left form page.'),
```
Comment (anthony-malkoun) [inline modules/tide_edit_protection/src/Form/TideEditProtectionForm.php:99]: "Choose which entity forms will be matched for unsaved changes warning when leaving form page.

File: modules/tide_edit_protection/src/Form/TideEditProtectionForm.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+          ':input[name="entity_method"]' => ['value' => 'entity_bundle'],
+        ],
+      ],
+    ];
+
+    $form['alert_form_ids'] = [
+      '#type' => 'textarea',
+      '#title' => $this->t('Other form ids'),
+      '#default_value' => $config->get('alert_form_ids'),
+      '#description' => $this->t('Specify forms by using their form id. Enter one form id per line. PHP regex [ tide_.*_form ] is support.'),
```
Comment (anthony-malkoun) [inline modules/tide_edit_protection/src/Form/TideEditProtectionForm.php:132]: "supported"

**Review by MdNadimHossain — COMMENTED** [review]
@vincent-gao do you think making it an independent module will give more flexibility to maintain this module?
Is there a specific reason this needs to be a module that belongs to tide_core?

File: modules/tide_edit_protection/tide_edit_protection.module [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * Implements hook_help().
+ */
+function tide_edit_protection_help($route_name, RouteMatchInterface $route_match) {
+  switch ($route_name) {
+    case 'help.page.tide_edit_protection':
+      $output = '';
+      $output .= '<h3>' . t('About') . '</h3>';
+      $output .= '<p>' . t('Protects unsaved change.') . '</p>';
```
Comment (GROwen) [inline modules/tide_edit_protection/tide_edit_protection.module:21]:
```suggestion
      $output .= '<p>' . t('Protects unsaved changes.') . '</p>';
```

File: modules/tide_edit_protection/tide_edit_protection.module [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function tide_edit_protection_form_views_exposed_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (\Drupal::moduleHandler()->moduleExists('tide_site_restriction')) {
+    $form_ids = [
+      'views-exposed-form-tide-locked-content-page-1' => 'uid_current',
+    ];
+    foreach ($form_ids as $form_id => $filter) {
```
Comment (GROwen) [inline modules/tide_edit_protection/tide_edit_protection.module:98]: Why is a loop necessary if the array that's being passed to it is singular?

File: modules/tide_edit_protection/tide_edit_protection.links.menu.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+  title: Tide Edit Protection settings
+  description: 'Tide Edit Protection settings.'
+  route_name: tide_edit_protection.admin
+  parent: system.admin_config_content
+
+tide_edit_protection.view:
+  title: 'Locked content'
+  route_name: view.tide_locked_content.page_1
+  parent: system.admin_content
+  weight: 100
```
Comment (GROwen) [inline modules/tide_edit_protection/tide_edit_protection.links.menu.yml:11]: newline is needed at end-of-file.

File: modules/tide_edit_protection/src/Form/TideEditProtectionForm.php [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    return new static($container->get('config.factory'), $container->get('entity_type.manager'), $container->get('entity_type.bundle.info'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('tide_edit_protection.form');
+    $entity_types = $entity_bundles = [];
+    $_entity_types = $this->entityTypeManger->getDefinitions();
```
Comment (GROwen) [inline modules/tide_edit_protection/src/Form/TideEditProtectionForm.php:67]: This naming convention is a little confusing. Can you improve the specificity of the variable and remove the prefix `_`? Same for `$_entity_bundles`

File: modules/tide_edit_protection/tide_edit_protection.module [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        $entity_bundles = $config->get('entity_bundles');
+        if (in_array($entity_bundle, $entity_bundles)) {
+          $include = TRUE;
+        }
+      }
+    }
+  }
+
+  // Check other forms by form_id, support Regex.
+  if (FALSE === $include) {
```
Comment (GROwen) [inline modules/tide_edit_protection/tide_edit_protection.module:62]: Has there been a team decision to use Yoda conditionals? I think SDP's "unofficial" (a discussion in Slack in a channel far, far away) code style preference is not to use them.

**Review by GROwen — CHANGES_REQUESTED** [review]

File: modules/tide_edit_protection/tide_edit_protection.module (hunk 5 above)
Comment (vincent-gao) [reply modules/tide_edit_protection/tide_edit_protection.module:98]: Thanks @GROwen , yes. removed the loop.

File: modules/tide_edit_protection/tide_edit_protection.module (hunk 8 above)
Comment (vincent-gao) [reply modules/tide_edit_protection/tide_edit_protection.module:62]: Thanks @GROwen . oh~ sorry, forgot about this . removed.

File: modules/tide_edit_protection/tide_edit_protection.module (hunk 4 above)
Comment (vincent-gao) [reply modules/tide_edit_protection/tide_edit_protection.module:21]: Thanks @GROwen , changed.

File: modules/tide_edit_protection/tide_edit_protection.links.menu.yml (hunk 6 above)
Comment (vincent-gao) [reply modules/tide_edit_protection/tide_edit_protection.links.menu.yml:11]: Thanks @GROwen , Changed.

File: modules/tide_edit_protection/src/Form/TideEditProtectionForm.php (hunk 7 above)
Comment (vincent-gao) [reply modules/tide_edit_protection/src/Form/TideEditProtectionForm.php:67]: Thanks @GROwen , changed.

**Comment by vincent-gao** [comment]
> @vincent-gao do you think making it an independent module will give more flexibility to maintain this module?
> Is there a specific reason this needs to be a module that belongs to tide_core?

Thanks @MdNadimHossain , no, there is no specific reason.~

**Comment by vincent-gao** [comment]
Hi @GROwen / @MdNadimHossain , just finished off all required changes. please take a look. thanks

**Comment by vincent-gao** [comment]
Hey @anthony-malkoun could you please take a look~

File: modules/tide_edit_protection/src/Form/TideEditProtectionForm.php (hunk 7 above)
Comment (anthony-malkoun) [reply modules/tide_edit_protection/src/Form/TideEditProtectionForm.php]: To confirm, no Yoda conditionals. Thanks for changing.

File: modules/tide_edit_protection/tide_edit_protection.module (hunk 8 above)
Comment (anthony-malkoun) [reply modules/tide_edit_protection/tide_edit_protection.module]: To confirm, no Yoda conditionals. Thanks for changing.

**Review by anthony-malkoun — APPROVED** [review]
LGTM. Thank you for making the recommended changes @vincent-gao.

**Pull request merged**